src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -1464,127 +1660,85 @@ public static void Sort(Array keys, Array? items, int index, int length, ICompar
             if (keys.Length - (index - keysLowerBound) < length || (items != null && (index - keysLowerBound) > items.Length - length))
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
 
-            if (length > 1)
+            if (length <= 1)
+                return;
+
+            comparer ??= Comparer.Default;
+
+            if (keys is object[] objKeys)
             {
-                comparer ??= Comparer.Default;
+                object[]? objItems = items as object[];
+                if (items == null || objItems != null)
+                {
+                    new SorterObjectArray(objKeys, objItems, comparer).Sort(index, length);
+                    return;
+                }
+            }
```

Comment:
I'm curious why you moved this up front?

Reply:
`object[]` is fast check that does not involve reflection. I though that it makes sense to put it first before anything that involves reflection. But I may need to rethink how these switches are done. The original implementation had special casing for IntPtr/UIntPtr that I have lost with this change...